packages/compose/src/hooks/use-event/index.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useInsertionEffect, useCallback } from '@wordpress/element';
+
+/**
+ * Any function.
+ */
+export type AnyFunction = ( ...args: any ) => any;
+
+/**
+ * Creates a stable callback function that has access to the latest state and
+ * can be used within event handlers and effect callbacks. Throws when used in
+ * the render phase.
+ *
+ * @param callback The callback function to wrap.
+ *
+ * @example
+ *
+ * ```tsx
+ * function Component(props) {
+ *   const onClick = useEvent(props.onClick);
+ *   React.useEffect(() => {}, [onClick]);
+ * }
+ * ```
+ */
+export default function useEvent< T extends AnyFunction >(
+	/**
+	 * The callback function to wrap.
+	 */
+	callback?: T
+) {
+	const ref = useRef< AnyFunction | undefined >( () => {
+		throw new Error( 'Cannot call an event handler while rendering.' );
+	} );
+	useInsertionEffect( () => {
+		ref.current = callback;
+	} );
+	return useCallback< AnyFunction >(
+		( ...args ) => ref.current?.( ...args ),
+		[]
+	) as T;
```

Comment:
I've seen folks putting thought into optimizing this extra by using `useRef` instead of `useCallback` to avoid creating the empty dependency array on every render:

https://github.com/scottrippey/react-use-event-hook/blob/9befe8c77ee97f9ce722b16ff7686822a98b25e1/src/useEvent.ts#L28-L37

Maybe that's premature optimization, but just wanted to bring it up anyway.

Reply:
Interesting, but it does seem like premature optimization to me. Every hook with deps is "creating" an array on every single render but I've never heard of that being a problem.

Reply:
Right.

Reply:
FWIW they never explicitly mentioned which array they're referring to, but there's no other array than the dependency array 🤔

Reply:
Taking a second look at the linked code, that is not concurrent-safe according to the rules. It might still work, but reading or writing a ref at render time is forbidden.

Reply:
Thought the same, and I thought I'm probably missing something by looking at the code. Thanks for confirming both!